memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+runtime must:
+
+1. If necessary, round up the size such that it's a multiple of the tag granule
+   size (i.e. ``size % 16 == 0``), and
+
+2. If necessary, round up the alignment such that it's a multiple of the tag
+   granule size (i.e. ``alignment % 16 == 0``).
+
+Static Linking
+--------------
+
+If the translation unit contains tagged global variables, the compiler should
+produce an ``SHT_AARCH64_MEMTAG_GLOBALS_STATIC`` section, as well as a
+relocation section with ``sh_link`` pointing to the
+``SHT_AARCH64_MEMTAG_GLOBALS_STATIC`` section.
+
+For each tagged global variable, the compiler should produce an ``R_NONE``
+relocation placed in the aforementioned relocation section. This relocation
+should have its ``ELF64_R_SYM`` bits of the ``r_info`` field point to the global
+variable that needs to be tagged. This communicates to the linker that the
+global variable in question is marked for tagging.
+
+In addition, the linker should ensure that all references to tagged global
+variable are via. the GOT (including for internal symbols). The relocations
+described later in this document ensure that the address tag for a global
+variable is materialized into the GOT, resulting in correctly tagged global
+variable accesses.
+
+Dynamic Linking
+---------------
+
+The dynamic linker should parse the ``SHT_AARCH64_MEMTAG_GLOBALS_STATIC``
+section to identify all tagged global variables.
+
+Resolving Multiple Definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Ideally, all translation units linked into a program are all built with or
+without tagged globals. This may not always be possible however, and so the
+linker must take special care when resolving the definition of a global variable
+if there are duplicate definitions that are not all tagged or untagged. In these
+instances, the linker must:
+
+1. Ensure the resolved definition is untagged, and
+
+2. Use the largest size and alignment for the resolved definition.
+
+Because tagged globals may have a larger size or alignment (in order to align
+with the tag granularity), the compiler may take advantage of optimisations that
+are only possible with the larger size or alignment. These optimisations may
+result in out-of-bounds memory accesses or alignment-related exceptions if the
+smaller size/alignment definition is selected.
+
+Encoding of ``SHT_AARCH64_MEMTAG_GLOBALS_DYNAMIC``
+--------------------------------------------------
+
+Each tagged global variable should have a random tag assigned by the dynamic
+loader at load time. In order to accomplish this, the loader needs to know the
+set of regions that need tagging (exactly which symbol corresponds to each
+region is not necessary). For each global variable, we emit a compressed
+"descriptor" to the ``SHT_AARCH64_MEMTAG_GLOBALS_DYNAMIC`` section.
+
+The descriptor is comprised of two parts:
+
+1. The unrelocated virtual address of the global variable, and
+
+2. The size of the global variable.
```

Comment:
This is the granule boundary aligned/extended size, correct? Would be good to make that 100% clear, lest someone think this is sizeof(the variable).

Actually I see below you talk about the size being granule aligned. Though I think it would still help to note that here.